### PR TITLE
Allow hosts to remain partially up (JAVA-544).

### PR DESCRIPTION
**Pull request by olim7t**
When a connection fails, don't mark the host down immediately: keep going with the remaining connections, and periodically try to reopen the missing ones. Only mark the host down when the last connection goes down.

This expands `ConvictionPolicy` to track the number of connections per host.

Missing connections are reopened by the connection pool. This means it requires activity (borrow / return), so a completely inactive pool might not recreate its connections, even stay below its core size. I think that's a reasonable tradeoff, given that a separate reconnection process à la `AbstractReconnectionHandler` would add a lot of complexity.

On a side note, I've introduced a dedicated `Host.statesLogger`, in an attempt to centralize all logs related to host states.

### CODE REVIEW

**Comment by olim7t** [comment]
The Clirr failures are all false positives, I will fix them shortly.

**Comment by zznate** [comment]
Is this going to integrate with the work in JAVA-720 at all? I always wanted to add 'coordinator or replica issue?' weighting to conviction policies in Hector to have more intelligent bad host take down semantics, but never got around to it. 

Either way, I have a bunch of accumulated thoughts here. I'll try to pull them together on the ticket after a deeper look and leave something more thoughtful/actionable :) 

Super cool to see this level of detail coming in though!

**Comment by olim7t** [comment]
> bad host take down semantics

I'm not sure if taking the host down is the best response. If the node is behaving badly but still accepting connections (for example a node that is isolated from other replicas but not the driver), the driver will end up reconnecting immediately.
Another approach that was suggested to us is to have LatencyAwarePolicy penalize hosts that throw errors (see [JAVA-837](https://datastax-oss.atlassian.net/browse/JAVA-837)).

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 1]
```diff
@@ -120,31 +124,33 @@ public void run() {
...
-        ListenableFuture<List<Void>> allConnectionsFuture = Futures.allAsList(connectionFutures);
+        ListenableFuture<List<Void>> allConnectionsFuture = Futures.successfulAsList(connectionFutures);
 
         final SettableFuture<Void> initFuture = SettableFuture.create();
-        Futures.addCallback(allConnectionsFuture, new FutureCallback<List<Void>>() {
+        Futures.addCallback(allConnectionsFuture, new MoreFutures.SuccessCallback<List<Void>>() {
             @Override
             public void onSuccess(List<Void> l) {
+                // Some of the connections might have failed, keep only the successful ones
+                ListIterator<Connection> it = connections.listIterator();
```
Comment (newkek) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Isn't `successfulAsList()` supposed to return only successful connections ? If so why would there be failed connections in `allConnectionsFuture` ?

Comment (newkek) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Or `successfulAsList` returns successful `ListenableFuture`s but the connection might have failed

Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: `connections` is not the result of the "compound" future, it's a list we built previously with all the connections, before initializing them (see lines 107, 121).

The result of `successfulAsList` is what gets passed to the callback (`List<Void> l`). It contains the results for the futures that have succeeded, and `null` for those that have failed. But since our futures are `Future<Void>`, the successful results are also `null`, so we can't do anything with `l`.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 2]
```diff
@@ -120,31 +124,33 @@ public void run() {
 
         Executor initExecutor = manager.cluster.manager.configuration.getPoolingOptions().getInitializationExecutor();
 
-        ListenableFuture<List<Void>> allConnectionsFuture = Futures.allAsList(connectionFutures);
+        ListenableFuture<List<Void>> allConnectionsFuture = Futures.successfulAsList(connectionFutures);
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:127]: One catch here is that if the _first_ connection we try to open fails, the host's connection count goes back to 0 and the host is marked down immediately.
However this only happen for the initial connection. For reconnection attempts, we know that a connection has been opened by the reconnection handler before attempting to create pools.
So I don't think we need to be too clever here. The alternative would be to wait until all connections have been attempted before marking down, which sounds too convoluted.

**Comment by olim7t** [comment]
I just opened [JAVA-856](https://datastax-oss.atlassian.net/browse/JAVA-856), but I think this PR will solve it.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 3]
```diff
@@ -475,17 +494,15 @@ private void maybeSpawnNewConnection() {
...
             open.decrementAndGet();
-        if (connections.remove(connection))
-            manager.blockingExecutor().submit(new Runnable() {
-                @Override
-                public void run() {
-                    addConnectionIfUnderMaximum();
-                }
-            });
-        connection.closeAsync();
+        connections.remove(connection);
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:506]: Shouldn't we call `closeAsync()` anymore?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 4]
```diff
@@ -168,16 +174,20 @@ public Connection borrowConnection(long timeout, TimeUnit unit) throws Connectio
...
         if (connections.isEmpty()) {
-            for (int i = 0; i < options().getCoreConnectionsPerHost(hostDistance); i++) {
-                // We don't respect MAX_SIMULTANEOUS_CREATION here because it's  only to
-                // protect against creating connection in excess of core too quickly
-                scheduledForCreation.incrementAndGet();
-                manager.blockingExecutor().submit(newConnectionTask);
+            if (!host.convictionPolicy.canReconnectNow())
+                throw new TimeoutException("Connection pool is empty, currently trying to reestablish connections");
+            else {
+                for (int i = 0; i < options().getCoreConnectionsPerHost(hostDistance); i++) {
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: It looks like with the DefaultConvictionPolicy isn't it possible that a reconnect might be in process while creating connections here?  Could that cause us to go above max simultaneous connections?

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java]: Answering my own question: **No** that can't happen since `newConnectionTask` ultimately calls `addConnectionIfUnderMaximum` which bails out if we are at or above maximum connections per hot.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 5]
```diff
@@ -464,6 +480,9 @@ else if (chosen.state.compareAndSet(TRASHED, RESURRECTING))
     }
 
     private void maybeSpawnNewConnection() {
+        if (!host.convictionPolicy.canReconnectNow())
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:488]: What if we have been naturally growing/shrinking connections?  Spawning seems like a good idea if connections aren't being closed for bad reasons.

File: driver-core/src/test/java/com/datastax/driver/core/AbstractReconnectionHandlerTest.java [hunk 6]
```diff
@@ -330,7 +330,7 @@ protected void onReconnection(Connection connection) {
     private void waitForCompletion() {
         executor.shutdown();
         try {
-            boolean shutdown = executor.awaitTermination(10, TimeUnit.SECONDS);
+            boolean shutdown = executor.awaitTermination(30, TimeUnit.SECONDS);
```
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/AbstractReconnectionHandlerTest.java:333]: why the change from 10 to 30?

File: driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java [hunk 7]
```diff
@@ -50,30 +66,89 @@
...
-            return true;
+        private synchronized void resetReconnectionTime() {
+            reconnectionSchedule = null;
+            nextReconnectionTime = Long.MIN_VALUE;
         }
 
         @Override
-        public void reset() {}
+        boolean canReconnectNow() {
+            return System.currentTimeMillis() >= nextReconnectionTime;
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java]: It seems weird to me for the ConvictionPolicy to dictate when to reconnect.  I think it would be much easier to manage if it was only responsible for whether or not to mark the connection down, especially if we ever intend to expose this.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 8]
```diff
@@ -221,7 +231,9 @@ public Connection borrowConnection(long timeout, TimeUnit unit) throws Connectio
         }
 
         int connectionCount = open.get() + scheduledForCreation.get();
-        if (connectionCount < options().getMaxConnectionsPerHost(hostDistance)) {
+        if (connectionCount < options().getCoreConnectionsPerHost(hostDistance)) {
+            maybeSpawnNewConnection();
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:240]: Just trying to understand: earlier in this method you say that "We don't respect MAX_SIMULTANEOUS_CREATION" when # connections < core connections and indeed we submit newConnectionTask right away; why is it different here?

File: driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java (hunk 7 above)
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java]: Agreed, `canReconnectNow()` seems to logically belong to `ReconnectionPolicy`.
On a side note, the purpose of `updateReconnectionTime` and `canReconnectNow` is to measure elapsed time, so wouldn't `System.nanoTime()` be a better choice here (think leap seconds, clock drift etc.)?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 3 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:506]: The connection is defunct. `defunct()` calls `closeAsync()`.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 5 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:488]: If `! canReconnectNow()`, it means connections were closed for bad reasons a short while ago.

File: driver-core/src/test/java/com/datastax/driver/core/AbstractReconnectionHandlerTest.java (hunk 6 above)
Comment (olim7t) [inline driver-core/src/test/java/com/datastax/driver/core/AbstractReconnectionHandlerTest.java:333]: I got a timeout while running the test locally.

File: driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java (hunk 7 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java]: We _could_ extract `updateReconnectionTime`, `resetReconnectionTime` and `canReconnectNow` somewhere else, but I'm not convinced. IMO the real issue is that `ConvictionPolicy` should probably be renamed. But it's about collecting failures and deciding when you can reconnect based on past failures, so it doesn't shock me that both responsibilities go in the same class.

`ReconnectionPolicy` only manages the reconnection schedule, not what we do with them. Plus changing it would break binary compatibility.

I don't think we should expose `ConvictionPolicy`. In fact I'm tempted to get rid of the abstract class and factory. As noted by Sylvain in the comments, "maybe just exposing say a threshold of error before convicting a node is enough".

> wouldn't System.nanoTime() be a better choice

Agreed.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:240]: It's done in `maybeSpawnNewConnection`.

File: driver-core/src/test/java/com/datastax/driver/core/AbstractReconnectionHandlerTest.java (hunk 6 above)
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/AbstractReconnectionHandlerTest.java:333]: I've noticed this too (on other branches not just this one).  Would you be averse to moving these tests from 'unit' to 'short'?

File: driver-core/src/main/java/com/datastax/driver/core/Connection.java [hunk 9]
```diff
@@ -581,9 +583,9 @@ boolean setPool(HostConnectionPool pool) {
      * The connection should generally not be reused after that.
      */
     void release() {
-        HostConnectionPool pool = poolRef.get();
-        if (pool != null)
-            pool.returnConnection(this);
+        Owner owner = ownerRef.get();
+        if (owner instanceof HostConnectionPool)
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Connection.java:588]: Found this in testing and think this may be an issue:  This only releases the Connection once, however even if the connection had many requests inFlight the host would believe it would have more requests inFlight then it actually does.

Consider the following case:
1. Create a Pool with core(1), max(2), maxSimultaneousRequests(100).
2. Acquire 101 connections on core, new connection 'extra1' will be created.  totalInFlight=101
3. Close core connection, it is released, totalInFlight = 100.  'extra1' becomes core.
4. Attempt to borrow a connection after reconnect interval.  Since totalInFlight=100 a new connection is spawned even though it shouldn't be.

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Connection.java:588]: Hrmm, maybe I'm wrong about how this goes about, but I do see that inFlight is not reduced on connection close.  However, returnConnection would eventually be called on request timeout/cancel, so maybe this isn't a big deal.

File: driver-core/src/test/java/com/datastax/driver/core/AbstractReconnectionHandlerTest.java (hunk 6 above)
Comment (olim7t) [inline driver-core/src/test/java/com/datastax/driver/core/AbstractReconnectionHandlerTest.java:333]: No, I agree we can move them to 'short'.

File: driver-core/src/main/java/com/datastax/driver/core/Connection.java (hunk 9 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Connection.java:588]: It sounds like you tried this scenario with the harness in `HostConnectionPoolTest`. The problem is that this harness is not well-suited to test for errors: in real life, if a connection gets closed, the way inFlight gets decreased is by failing all pending `ResponseCallbacks`, which will themselves release the connection (see `SpeculativeExecution#onException`).
In `HostConnectionPoolTest` there are no handlers, we're just borrowing and releasing connections manually. In fact I'm not even sure why totalInFlight was decremented.
One thing we could do in the test is use a simple `ResponseCallback` implementation instead of borrowing and releasing manually.

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Connection.java:588]: :+1:, I'm using your 'MockRequest' implementation and it's working great, thanks!

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:240]: This would only spawn 1 new connection, not up to core connections.  If I have a scenario where I have a Core Pool Size for 8 and 4 of the 8 connections are established successfully on pool init, this will spawn 1 new connection instead of 4.  However, if I were to make 4 requests it would get me up to 8 connections since it will spawn a new connection on every borrow up to core, so to me that doesn't seem like we need any special logic for this, since if we don't end up making up to core pool size requests on a Host, we don't need up to core pool size connections!

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 1 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: I've noticed that if all connections fail the host is not marked down.  This is because ConvictionPolicy does not mark host down if connection was initializing.  This leaves us with an empty set of connections in the success callback here.    It will try to bring connections up to core properly, but I think you could argue in this case the host could be marked down.  What do you think?  I can see it both ways.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 2 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:127]: I think your change in DefaultConvictionPolicy#signalConnectionFailure to not consider host down if connectionInitialized is false gets around this though right?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 1 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Hrmm, though I could see that in this case we should treat this the same as having 0 core connections (we'll try connecting if we can, leave the control connection in tact if it stays up and heartbeats are working)...so maybe we should leave this as is.

File: driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java [hunk 10]
```diff
@@ -50,30 +70,90 @@
...
         }
 
-        public boolean addFailureFromExternalDetector() {
-            return true;
+        @Override
+        void signalConnectionClosed() {
+            if (host.state == Host.State.DOWN)
+                return;
+            int remaining = openConnections.decrementAndGet();
+            assert remaining >= 0;
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java:100]: I noticed a valid case for this:   If someone calls Cluster.close() it ultimately calls Connection.closeAsync() which calls signalConnectionClosed(), depending on ordering of connections closing you may sometimes get this exception:

```
java.lang.AssertionError
    at com.datastax.driver.core.ConvictionPolicy$DefaultConvictionPolicy.signalConnectionClosed(ConvictionPolicy.java:101)
    at com.datastax.driver.core.Connection.closeAsync(Connection.java:622)
    at com.datastax.driver.core.HostConnectionPool.discardAvailableConnections(HostConnectionPool.java:607)
    at com.datastax.driver.core.HostConnectionPool.closeAsync(HostConnectionPool.java:585)
    at com.datastax.driver.core.SessionManager.closeAsync(SessionManager.java:116)
    at com.datastax.driver.core.Cluster$Manager.close(Cluster.java:1513)
    at com.datastax.driver.core.Cluster$Manager.access$200(Cluster.java:1199)
    at com.datastax.driver.core.Cluster.closeAsync(Cluster.java:490)
    at com.datastax.driver.core.Cluster.close(Cluster.java:501)
    at com.datastax.driver.core.HostConnectionPoolTest.should_not_close_trashed_connection_until_no_in_flight(HostConnectionPoolTest.java:335)
```

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java:100]: I think for this to happen you need to have a Connection with pending requests sitting in the trash.  Noticed this in my logs slightly before:

```
    41927  [cluster2-scheduled-task-worker-0] DEBUG com.datastax.driver.core.Connection  - Not terminating Connection[/127.0.1.1:9042-2, inFlight=0, closed=true]: there are still pending requests
```

**Note**:  This was a bit of a bug in my code, I'm returning on a connection but not emptying the dispatcher when I return the connection, but it creates this scenario which I thought was interesting.

It triggers a heartbeat on that connection which is closed and in the trash:

```
    61944  [cluster2-nio-worker-1] DEBUG com.datastax.driver.core.Connection  - Connection[/127.0.1.1:9042-2, inFlight=0, closed=true] was inactive for 30 seconds, sending heartbeat
    61944  [cluster2-nio-worker-1] DEBUG com.datastax.driver.core.Connection  - Not terminating Connection[/127.0.1.1:9042-2, inFlight=0, closed=true]: there are still pending requests
    61944  [cluster2-nio-worker-1] DEBUG com.datastax.driver.core.Connection  - Connection[/127.0.1.1:9042-2, inFlight=0, closed=true] connection error
com.datastax.driver.core.ConnectionException: [/127.0.1.1:9042] Connection has been closed
    at com.datastax.driver.core.Connection.write(Connection.java:520)
    at com.datastax.driver.core.Connection.write(Connection.java:496)
    at com.datastax.driver.core.Connection$Dispatcher.userEventTriggered(Connection.java:1000)
    at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:308)
    at io.netty.channel.AbstractChannelHandlerContext.fireUserEventTriggered(AbstractChannelHandlerContext.java:294)
    at io.netty.handler.timeout.IdleStateHandler.channelIdle(IdleStateHandler.java:322)
    at io.netty.handler.timeout.IdleStateHandler$AllIdleTimeoutTask.run(IdleStateHandler.java:435)
    at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
    at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120)
    at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:357)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:111)
    at java.lang.Thread.run(Thread.java:745)
```

This signals a failure on the connection (I think) which decrements the count.

If I add !isClosed to the condition check in userEventTriggered in Connection$Dispatcher the error goes away since we don't try sending a heartbeat on a closed connection in the trash.  Also fixing my problem (not emptying the dispatcher when returning a connection) fixed this as well.  I think it would be a good idea to do a 'isClosed' check before sending a heartbeat anyways.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 1 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Playing with this some more and I think maybe we do need to change the behavior, consider this scenario:   You have a 2 node cluster with node1 being up and running and node 2 being down.   You create a Cluster object with node1 being your contact point.   node2 is discovered in system.peers so we create a pool to it.  All connections fail, yet the host remains in the 'UP' state.  Wouldn't that be a problem?  We'd keep trying to borrow connections on node2 and would get exceptions (though would move on to the next host).

A test that effectively does this:  RecommissionedNodeTest#should_ignore_recommissioned_node_on_control_connection_reconnect.

File: driver-core/src/main/java/com/datastax/driver/core/Connection.java [hunk 11]
```diff
@@ -196,6 +196,9 @@ public void operationComplete(ChannelFuture future) throws Exception {
         Futures.addCallback(initFuture, new FutureCallback<Void>() {
             @Override
             public void onSuccess(Void result) {
+                Host host = factory.manager.metadata.getHost(address);
+                if(host != null)
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Connection.java:200]: Added a null check here and also in ef78e63 as I saw this raised in a test that decommissioned hosts.  Not 100% sure if this is the most appropriate fix.  I'm guessing this happens for connections belonging to a ControlConnection that had a removed Host?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 8 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:240]: Yes, that was the intent. Here we know that we already have a connection to execute the current request, so we don't need to be overly aggressive in restoring the core size. As you noted it will be reached over time.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 1 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Yes, that's an oversight, the node should be marked down if all connections fail to initialize. Will look into it.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 2 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:127]: That's correct. I'll re-check the initialization scenario based on your other remarks.

File: driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java (hunk 10 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java:100]: Yes, checking for isClosed makes sense, adding the check.

File: driver-core/src/main/java/com/datastax/driver/core/Connection.java (hunk 11 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Connection.java:200]: This happens when the connection succeeds after the node was concurrently removed. I think we need to understand the scenario better to find out what fix is most appropriate.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 2 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:127]: The goal of `wasConnectionInitialized` in the conviction policy was really to avoid decrementing the count.
We could still mark the node down if the count is 0. The only problem is that we would signal multiple times if all pooled connections fail to initialize, but the code already handles that (in fact, that's what happens with the current version).

File: driver-core/src/main/java/com/datastax/driver/core/Connection.java (hunk 11 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Connection.java:200]: :+1: i'll look more into both the connection success/connection closed scenarios where the Host is gone and let you know what I find.

File: driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java (hunk 10 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java:100]: That looks great thanks

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 1 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Makes sense, with your change a few of the tests that were failing previously because nodes weren't being marked down are now passing (the one I mentioned being one of them), but there are still some remaining (mostly in RecommissionedNodeTest).  Will look into this.

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: It looks like this works in all cases on initialization except when we get an UnsupportedProtocolVersion/ClusterNameMismatchException because we don't defunct the connection in these cases (and thus don't call signalConnectionFailure in those cases), so the host remains up:

``` java
        // Fallback on initializeTransportFuture so we can properly propagate specific exceptions.
        ListenableFuture<Void> initFuture = Futures.withFallback(initializeTransportFuture, new FutureFallback<Void>() {
            @Override
            public ListenableFuture<Void> create(Throwable t) throws Exception {
                SettableFuture<Void> future = SettableFuture.create();
                // Make sure the connection gets properly closed.
                if (t instanceof ClusterNameMismatchException || t instanceof UnsupportedProtocolVersionException) {
                    // These exceptions cause the node to be ignored, so just propagate
                    closeAsync().force();
                    future.setException(t);
                } else {
                    // Defunct to ensure that the error will be signaled (marking the host down)
                    Exception e = (t instanceof ConnectionException || t instanceof DriverException || t instanceof InterruptedException)
                        ? (Exception)t
                        : new ConnectionException(Connection.this.address,
                        String.format("Unexpected error during transport initialization (%s)", t),
                        t);
                    future.setException(defunct(e));
                }
                return future;
            }
        }, initExecutor);
```

If I add signalConnectionFailure in the if branch for ClusterNameMismatch/UnsupportedProtocolVersion the host is now marked down, but not sure if that is the appropriate fix.

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: It looks like this _shouldn't_ be a problem if we get a connection timeout because the Futures.withFallback will cover channelReadyFuture failing, but thought I'd point out that scenario as well.

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Hrmm, another problem here is that even if we signalConnectionDown this will trigger a reconnect for Unsupported Protocol Version which we test for (RecommissionedNodeTest#should_ignore_node_that_does_not_support_protocol_version_on_session_init) so there might be some logic issue somewhere that I haven't tracked down yet.

Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Yes, ClusterNameMismatch/UnsupportedProtocolVersion should be bubbled up to `SessionManager#maybeAddPool`, which will handle them accordingly (mark the node down and don't trigger a reconnect).
I'm fixing this on my local integration branch because it's easier to do on top of 821/822 (very similar to handling a 'wrong keyspace' error).
Can you keep a note to retry these tests with 822?

Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: PS - `should_ignore_node_that_does_not_support_protocol_version_on_session_init` works with my local fix.

Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:134]: Will do, this was the last thing that was giving my problems so I think this branch is good to merge.  I still need to investigate the null host thing, but I think I can look into that on the 822 branch after it is rebased.   I think this branch is in good enough shape to merge now if it's ok with you.   :+1:

**Comment by olim7t** [comment]
Merged with a build error (`ClusterInitTest#should_be_able_to_close_cluster_that_never_successfully_connected`).
It works on my machine and I can't find any way to relaunch the Travis build.
